Slepc 3.6.3 compatibility

With slepc-3.6.3 + petsc-3.6.4 freshly installed with candi I get the following:

~~~
/usr/tmp/heltai/dealii/build/lib/libdeal_II.g.so.8.5.0-pre: error: undefined reference to 'EPSComputeResidualNorm'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
~~~

anyone else?

The documentation of slepc seems to refer to a different interface. Apparently they maintained the definition but not the implementation...

Thanks Luca, I will check it out tomorrow morning. I am using petsc/slepc
dev without this peroblem. I may have missed something.

Best;
   Toby


21.03.2017 17:57 "Luca Heltai" <notifications@github.com> napisał(a):

With slepc-3.6.3 + petsc-3.6.4 freshly installed with candi I get the
following:

/usr/tmp/heltai/dealii/build/lib/libdeal_II.g.so.8.5.0-pre: error:
undefined reference to 'EPSComputeResidualNorm'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

anyone else?

The documentation of slepc seems to refer to a different interface.
Apparently they maintained the definition but not the implementation...

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/dealii/dealii/issues/4089>, or mute the thread
<https://github.com/notifications/unsubscribe-auth/AFU33naFzfPiHrs5EBybIwZDnZF-X4qoks5roAGJgaJpZM4MkGyc>
.

> With slepc-3.6.3 + petsc-3.6.4 freshly installed with candi I get the
> following:
> 
> ~~~
> /usr/tmp/heltai/dealii/build/lib/libdeal_II.g.so.8.5.0-pre: error:
> undefined reference to 'EPSComputeResidualNorm' collect2: error: ld
> returned 1 exit status ninja: build stopped: subcommand failed.
> ~~~
> 
> anyone else?
> 

My deal.II is building with the same petsc/slepc version numbers you
quoted (candi free). That will take a while. ;-)

> The documentation of slepc seems to refer to a different interface.
> Apparently they maintained the definition but not the
> implementation... 

Yes, the SLEPc manual states, "Interface changes:
EPSComputeRelativeError and EPSComputeResidualNorm have been deprecated
(use EPSComputeError instead)". 

I don't know how SLEPc "deprecates" functions, though I assume the
function implementation should still be there. Which means something is
broken. Does candi run the petsc and slepc tests? These pass for me.

(Not much help, sorry).

--- a/source/lac/slepc_solver.cc
+++ b/source/lac/slepc_solver.cc
@@ -195,22 +195,29 @@ namespace SLEPcWrappers
       for (unsigned int i = 0; i < *n_converged; i++)
         {
           double residual_norm_i = 0.0;
-          // EPSComputeResidualNorm is L2-norm and is not consistent with the stopping criteria
-          // used during the solution process.
-          // Yet, this is the norm which gives error bounds (Saad, 1992, ch3):
-          //   | \lambda - \widehat\lambda | <= ||r||_2
-          ierr = EPSComputeResidualNorm (eps, i, &residual_norm_i);
-
-          // EPSComputeRelativeError may not be consistent with the stopping criteria
-          // used during the solution process. Given EPS_CONV_ABS set above,
-          // this can be either the l2 norm or the mass-matrix induced norm
-          // when EPS_GHEP is set.
-          // ierr = EPSComputeRelativeError (solver_data->eps, i, &residual_norm_i);
-
-          // EPSGetErrorEstimate is consistent with the residual norm
-          // used during the solution process. However, it is not guaranteed to
-          // be derived from the residual even when EPSSetTrueResidual is set.
-          // ierr = EPSGetErrorEstimate (solver_data->eps, i, &residual_norm_i);
+          // EPSComputeError (or, in older versions of SLEPc,
+          // EPSComputeResidualNorm) uses an L2-norm and is not consistent
+          // with the stopping criterion used during the solution process (see
+          // the SLEPC manual, section 2.5). However, the norm that gives error
+          // bounds (Saad, 1992, ch3) is (for Hermitian problems)
+          // | \lambda - \widehat\lambda | <= ||r||_2
+          //
+          // Similarly, EPSComputeRelativeError may not be consistent with the
+          // stopping criterion used in the solution process.
+          //
+          // EPSGetErrorEstimate is (according to the SLEPc manual) consistent
+          // with the residual norm used during the solution process. However,
+          // it is not guaranteed to be derived from the residual even when
+          // EPSSetTrueResidual is set: see the discussion in the thread
+          //
+          // https://lists.mcs.anl.gov/pipermail/petsc-users/2014-November/023509.html
+          //
+          // for more information.
+#if DEAL_II_SLEPC_VERSION_GTE(3, 6, 0)
+          ierr = EPSComputeError(eps, i, EPS_ERROR_ABSOLUTE, &residual_norm_i);
+#else
+          ierr = EPSComputeResidualNorm(eps, i, &residual_norm_i);
+#endif
 
           AssertThrow (ierr == 0, ExcSLEPcError(ierr));
           residual_norm = std::max (residual_norm, residual_norm_i);

